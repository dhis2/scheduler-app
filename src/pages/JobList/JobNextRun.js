import moment from 'moment'
import { PropTypes } from '@dhis2/prop-types'
import i18n from '@dhis2/d2-i18n'

const JobNextRun = ({ nextExecutionTime, enabled }) => {
    if (!enabled || !nextExecutionTime) {
        return '-'
    }

    const now = moment(Date.now())

    /**
     * The recommendation is to run dhis2 on a server set to UTC time.
     * In that case this timestamp is also UTC. If those recommendations
     * weren't followed the time could be off, but there's nothing
     * we can do to detect that.
     */
    const nextRun = moment.utc(nextExecutionTime)
    const nextRunIsInPast = nextRun.isSameOrBefore(now, 'minute')

    /**
     * If the time is in the past, that could mean that the task is running,
     * and the nextExecutionTime hasn't been updated yet.
     */

    if (nextRunIsInPast) {
        return i18n.t('Not scheduled')
    }

    return now.to(nextRun)
}

JobNextRun.propTypes = {
    enabled: PropTypes.bool.isRequired,
    nextExecutionTime: PropTypes.string,
}

export default JobNextRun
