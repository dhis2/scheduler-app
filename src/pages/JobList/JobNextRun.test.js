import React from 'react'
import { shallow } from 'enzyme'
import JobNextRun from './JobNextRun'

// Z is the zone designator for the zero UTC offset
const now = new Date('2010-10-10T10:10:10.000Z').valueOf()

// The server does not specify a timezone, but is assumed to provide UTC
const past = '2009-10-10T10:10:10.000'
const future = '2011-10-10T10:10:10.000'

describe('<JobNextRun>', () => {
    it('returns the next run time for an enabled job and a future execution time', () => {
        jest.spyOn(global.Date, 'now').mockImplementationOnce(() => now)

        const wrapper = shallow(
            <JobNextRun nextExecutionTime={future} enabled={true} />
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('returns fallback message for an enabled job and a past execution time', () => {
        jest.spyOn(global.Date, 'now').mockImplementationOnce(() => now)

        const wrapper = shallow(
            <JobNextRun nextExecutionTime={past} enabled={true} />
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('returns fallback message for a disabled job', () => {
        jest.spyOn(global.Date, 'now').mockImplementationOnce(() => now)

        const wrapper = shallow(
            <JobNextRun nextExecutionTime={''} enabled={false} />
        )

        expect(wrapper).toMatchSnapshot()
    })
})
