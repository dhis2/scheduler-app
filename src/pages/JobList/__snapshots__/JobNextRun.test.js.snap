// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JobNextRun> returns fallback message for a disabled job 1`] = `"-"`;

exports[`<JobNextRun> returns fallback message for an enabled job and a past execution time 1`] = `"Not scheduled"`;

exports[`<JobNextRun> returns the next run time for an enabled job and a future execution time 1`] = `"in a year"`;
